ygen/protogen.go

```diff
@@ -292,40 +296,162 @@ type protoMsgConfig struct {
+func protobufPackageForMsg(msg *yangDirectory, state *genState, compressPaths bool) (string, error) {
 	switch {
 	case msg.isFakeRoot:
 		// In this case, we explicitly leave the package name as nil, which is interpeted
 		// as meaning that the base package is used throughout the handling code.
+		return "", nil
 	case msg.entry.Parent == nil:
-		return nil, []error{fmt.Errorf("YANG schema element %s does not have a parent, protobuf messages are not generated for modules", msg.entry.Path())}
-	default:
-		// pkg is the name of the protobuf package, if the entry's parent has already
-		// been seen in the schema, the same package name as for siblings of this
-		// entry will be returned.
-		pkg = state.protobufPackage(msg.entry, cfg.compressPaths)
+		return "", fmt.Errorf("YANG schema element %s does not have a parent, protobuf messages are not generated for modules", msg.entry.Path())
+	}
+	// pkg is the name of the protobuf package, if the entry's parent has already
+	// been seen in the schema, the same package name as for siblings of this
+	// entry will be returned.
+	return state.protobufPackage(msg.entry, compressPaths), nil
+}
+
+// writeProto3MsgHierarchy generates a protobuf message definition. It takes the
+// arguments of writeProto3Message, outputting an individual message that outputs
+// a package definition and a single protobuf message.
+func writeProto3MsgSingleMsg(msg *yangDirectory, msgs map[string]*yangDirectory, state *genState, cfg protoMsgConfig) (*generatedProto3Message, []error) {
+	pkg, err := protobufPackageForMsg(msg, state, cfg.compressPaths)
+	if err != nil {
+		return nil, []error{err}
 	}
 
-	msgDefs, errs := genProto3Msg(msg, msgs, state, cfg, pkg)
+	msgDefs, errs := genProto3Msg(msg, msgs, state, cfg, pkg, nil)
 	if errs != nil {
 		return nil, errs
 	}
 
+	return genProto3MsgCode(pkg, msgDefs)
+}
+
+// genProto3MsgCode takes an input package name, and set of protobuf message
+// definitions, and outputs the generated code for the messages.
+func genProto3MsgCode(pkg string, msgDefs []protoMsg) (*generatedProto3Message, []error) {
 	var b bytes.Buffer
 	imports := map[string]interface{}{}
-	for _, msgDef := range msgDefs {
+	for i, msgDef := range msgDefs {
+		// Sort the child messages into a determinstic order. We cannot use the
+		// package name as a key as it may be the same for multiple packages, therefore
+		// use the code.
+		cmsgs := map[string]*generatedProto3Message{}
+		cstrs := []string{}
+		for _, m := range msgDef.ChildMsgs {
+			cmsgs[m.MessageCode] = m
+			cstrs = append(cstrs, m.MessageCode)
+		}
+		sort.Strings(cstrs)
+		nm := []*generatedProto3Message{}
+		for _, c := range cstrs {
+			nm = append(nm, cmsgs[c])
+		}
+		msgDef.ChildMsgs = nm
+
 		if err := protoTemplates["msg"].Execute(&b, msgDef); err != nil {
 			return nil, []error{err}
 		}
 		addNewKeys(imports, msgDef.Imports)
+		if i != len(msgDefs)-1 {
+			b.WriteRune('\n')
```

Comment:
Do this write outside the for loop without checking for value of i?

Reply:
I don't think we can do that, can we? Here we're wanting to append if the value is not the last value - so we need to do some check.

Reply:
yes, you'd still need to check for empty. i guess it's just a personal preference for being slightly easier to read, but your version is fine.

Reply:
Left this as is for now.